lib/bitcoin/store/utxo_db.rb

```diff
@@ -0,0 +1,254 @@
+require 'leveldb'
+
+module Bitcoin
+  module Store
+    class UtxoDB
+
+      KEY_PREFIX = {
+        out_point: 'o',        # key: out_point(tx_hash and index), value: Utxo
+        script: 's',           # key: script_pubkey and out_point(tx_hash and index), value: Utxo
+        height: 'h',           # key: block_height and out_point, value: Utxo
+        tx_hash: 't',          # key: tx_hash of transaction, value: [block_height, tx_index]
+        block: 'b',            # key: block_height and tx_index, value: tx_hash
+        tx_payload: 'p',       # key: tx_hash, value: Tx
+      }
+
+      attr_reader :level_db, :logger
+
+      def initialize(path = "#{Bitcoin.base_dir}/db/utxo")
+        FileUtils.mkdir_p(path)
+        @level_db = ::LevelDB::DB.new(path)
+        @logger = Bitcoin::Logger.create(:debug)
+      end
+
+      def close
+        level_db.close
+      end
+
+      def version
+        1
+      end
+
+      # @param  [String] hashes - array of String from hash160
+      #                         - can also be hash of redeem script
+      # @param  [Bitcoin::Tx] tx - incoming tx.
+      # @param  [Integer] block_height - block height number.
+      # @return [Bitcoin::OutPoint ...] 
+      def filter_and_save_tx(hashes, tx, block_height=nil)
```

Comment:
I do not know in what context or in which case this method should be called.

In my opinion,

(1) save_tx and save_utxo are called when the spv node receives a 'tx' message.

(2) save_tx_position is called when a 'merkleblock' or 'header' message is received.